File: src/csharp/Foundgine.Core/Semantic/SemanticGraph.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic;

/// <summary>
/// The canonical request graph. It contains only semantic/domain topology.
/// It must not contain SQL, GraphQL, provider nodes, aliases, or storage SQL.
/// A node may be reached through either a relational relationship or an AOT
/// semantic connection. A connection is a pre-resolved communication edge.
/// </summary>
public sealed class SemanticGraph
{
    private readonly List<SemanticGraphNode> _nodes = [];

    /// <summary>Creates an empty semantic graph for incremental construction and tests.</summary>
    public SemanticGraph()
    {
    }

    internal SemanticGraph(IEnumerable<SemanticGraphNode> nodes,
        Foundgine.Core.Semantic.Query.SemanticQueryOptions? options)
    {
        _nodes.AddRange(nodes);
        Options = options;
    }

    public IReadOnlyList<SemanticGraphNode> Nodes => _nodes;

    public Foundgine.Core.Semantic.Query.SemanticQueryOptions? Options { get; internal set; }

    /// <summary>
    /// Returns a semantic graph with an authorization predicate attached to an
    /// existing node. The graph topology and query options are preserved.
    /// This is the boundary used when policy is supplied after intent
    /// resolution but before planning.
    /// </summary>
    public SemanticGraph WithAuthorization(int nodeId, AuthorizationPredicate authorization)
    {
        ArgumentNullException.ThrowIfNull(authorization);

        if (_nodes.All(node => node.Id != nodeId))
            throw new ArgumentOutOfRangeException(nameof(nodeId));

        // Preserve node identities and topology. Only the affected immutable
        // record is replaced; this avoids rebuilding the entire graph.
        var nodes = _nodes
            .Select(node => node.Id == nodeId ? node with { Authorization = authorization } : node)
            .ToArray();

        return new SemanticGraph(nodes, Options);
    }

    public SemanticGraphNode AddRoot(
        EntityId entityId,
        IEnumerable<FieldId>? fields = null,
        AuthorizationPredicate? authorization = null) =>
        Add(entityId, null, null, null, fields, authorization);

    /// <summary>
    /// Adds a node reached through an AOT semantic connection. The connection
    /// identifies the pre-resolved communication edge; the target entity is
    /// already known and remains the only storage-side identity in the graph.
    /// </summary>
    public SemanticGraphNode AddConnection(
        EntityId entityId,
        ConnectionId connectionId,
        SemanticGraphNode parent,
        IEnumerable<FieldId>? fields = null,
        AuthorizationPredicate? authorization = null) =>
        Add(entityId, null, connectionId, parent, fields, authorization);

    public SemanticGraphNode Add(
        EntityId entityId,
        RelationshipId? relationshipId,
        SemanticGraphNode? parent,
        IEnumerable<FieldId>? fields = null) =>
        Add(entityId, relationshipId, parent, fields, null);

    /// <summary>
    /// Adds a node reached through a semantic relationship while preserving
    /// the AOT authorization predicate attached to that node.
    /// </summary>
    public SemanticGraphNode Add(
        EntityId entityId,
        RelationshipId? relationshipId,
        SemanticGraphNode? parent,
        IEnumerable<FieldId>? fields,
        AuthorizationPredicate? authorization) =>
        Add(entityId, relationshipId, null, parent, fields, authorization);

    private SemanticGraphNode Add(
        EntityId entityId,
        RelationshipId? relationshipId,
        ConnectionId? connectionId,
        SemanticGraphNode? parent,
        IEnumerable<FieldId>? fields = null,
        AuthorizationPredicate? authorization = null)
    {
        if (relationshipId is not null && connectionId is not null)
            throw new ArgumentException(
                "A semantic node cannot be reached through both a relationship and a connection.");

        if (parent is null && connectionId is not null)
            throw new ArgumentException("A root semantic node cannot be reached through a connection.");

        var node = new SemanticGraphNode(
            _nodes.Count,
            entityId,
            relationshipId,
            connectionId,
            parent?.Id,
            authorization)
        {
            Fields = fields?.Distinct().ToArray() ?? []
        };

        _nodes.Add(node);
        return node;
    }
}

public sealed record SemanticGraphNode(
    int Id,
    EntityId EntityId,
    RelationshipId? ViaRelationship,
    ConnectionId? ViaConnection,
    int? ParentId,
    AuthorizationPredicate? Authorization = null)
{
    /// <summary>Stable semantic annotations supplied by intent or semantic configuration.</summary>
    public IReadOnlyDictionary<string, string> SemanticAnnotations { get; init; } =
        new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);

    /// <summary>Provenance for how this node was traversed from the root.</summary>
    public SemanticTraversalOrigin? TraversalOrigin { get; init; }

    /// <summary>Provenance identifying the intent operation that created the node.</summary>
    public SemanticIntentOrigin? IntentOrigin { get; init; }

    /// <summary>Cardinality expected by the intent/planner at this node.</summary>
    public SemanticExpectedCardinality ExpectedCardinality { get; init; } = SemanticExpectedCardinality.Unknown;

    /// <summary>Whether the path can legally produce a null target.</summary>
    public bool IsNullablePath { get; init; }

    /// <summary>Provider-neutral semantic constraints relevant to this node.</summary>
    public IReadOnlyList<SemanticConstraint> SemanticConstraints { get; init; } = [];

    /// <summary>
    /// Fields selected for this entity in the request. These are semantic
    /// field identities only; no provider/storage information is carried.
    /// </summary>
    public IReadOnlyList<FieldId> Fields { get; init; } = [];
}

File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/SemanticGraph.java
package com.foundgine.core.semantic;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.query.SemanticQueryOptions;
import java.util.*;

/** Canonical request graph containing semantic/domain topology only. */
public final class SemanticGraph {
    private final List<SemanticGraphNode> nodes = new ArrayList<>();
    private SemanticQueryOptions options;

    public SemanticGraph() {}

    public SemanticGraph(Collection<SemanticGraphNode> nodes, SemanticQueryOptions options) {
        if (nodes != null) this.nodes.addAll(nodes);
        this.options = options;
    }

    public List<SemanticGraphNode> nodes() { return Collections.unmodifiableList(nodes); }
    public SemanticQueryOptions options() { return options; }
    public void setOptions(SemanticQueryOptions options) { this.options = options; }

    public SemanticGraph withAuthorization(int nodeId, AuthorizationPredicate authorization) {
        Objects.requireNonNull(authorization);
        if (nodes.stream().noneMatch(n -> n.id() == nodeId)) throw new IllegalArgumentException("Unknown node: " + nodeId);
        return new SemanticGraph(nodes.stream().map(n -> n.id() == nodeId ? n.withAuthorization(authorization) : n).toList(), options);
    }

    public SemanticGraphNode addRoot(EntityId entityId, Collection<FieldId> fields, AuthorizationPredicate authorization) {
        return add(entityId, null, null, null, fields, authorization);
    }
    public SemanticGraphNode addRoot(EntityId entityId) { return addRoot(entityId, List.of(), null); }
    public SemanticGraphNode addRoot(EntityId entityId, Collection<FieldId> fields) { return addRoot(entityId, fields, null); }

    public SemanticGraphNode addConnection(EntityId entityId, ConnectionId connectionId, SemanticGraphNode parent,
                                           Collection<FieldId> fields, AuthorizationPredicate authorization) {
        return add(entityId, null, connectionId, parent, fields, authorization);
    }
    public SemanticGraphNode addConnection(EntityId entityId, ConnectionId connectionId, SemanticGraphNode parent) {
        return addConnection(entityId, connectionId, parent, List.of(), null);
    }

    public SemanticGraphNode add(EntityId entityId, RelationshipId relationshipId, SemanticGraphNode parent,
                                 Collection<FieldId> fields) {
        return add(entityId, relationshipId, null, parent, fields, null);
    }
    public SemanticGraphNode add(EntityId entityId, RelationshipId relationshipId, SemanticGraphNode parent,
                                 Collection<FieldId> fields, AuthorizationPredicate authorization) {
        return add(entityId, relationshipId, null, parent, fields, authorization);
    }

    private SemanticGraphNode add(EntityId entityId, RelationshipId relationshipId, ConnectionId connectionId,
                                  SemanticGraphNode parent, Collection<FieldId> fields, AuthorizationPredicate authorization) {
        Objects.requireNonNull(entityId);
        if (relationshipId != null && connectionId != null) throw new IllegalArgumentException("A semantic node cannot be reached through both a relationship and a connection.");
        if (parent == null && connectionId != null) throw new IllegalArgumentException("A root semantic node cannot be reached through a connection.");
        var node = new SemanticGraphNode(nodes.size(), entityId, relationshipId, connectionId,
                parent == null ? null : parent.id(), authorization,
                fields == null ? List.of() : fields.stream().distinct().toList());
        nodes.add(node);
        return node;
    }

    public record SemanticGraphNode(
        int id, EntityId entityId, RelationshipId viaRelationship, ConnectionId viaConnection, Integer parentId,
        AuthorizationPredicate authorization, List<FieldId> fields, Map<String,String> semanticAnnotations,
        SemanticTraversalOrigin traversalOrigin, SemanticIntentOrigin intentOrigin,
        SemanticExpectedCardinality expectedCardinality, boolean nullablePath, List<SemanticConstraint> semanticConstraints) {
        public SemanticGraphNode {
            Objects.requireNonNull(entityId);
            fields = fields == null ? List.of() : List.copyOf(fields);
            semanticAnnotations = semanticAnnotations == null ? Map.of() : Collections.unmodifiableMap(new LinkedHashMap<>(semanticAnnotations));
            expectedCardinality = expectedCardinality == null ? SemanticExpectedCardinality.UNKNOWN : expectedCardinality;
            semanticConstraints = semanticConstraints == null ? List.of() : List.copyOf(semanticConstraints);
        }
        public SemanticGraphNode(int id, EntityId entityId, RelationshipId viaRelationship, ConnectionId viaConnection, Integer parentId, AuthorizationPredicate authorization) {
            this(id, entityId, viaRelationship, viaConnection, parentId, authorization, List.of(), Map.of(), null, null, SemanticExpectedCardinality.UNKNOWN, false, List.of());
        }
        public SemanticGraphNode(int id, EntityId entityId, RelationshipId viaRelationship, ConnectionId viaConnection, Integer parentId, AuthorizationPredicate authorization, List<FieldId> fields) {
            this(id, entityId, viaRelationship, viaConnection, parentId, authorization, fields, Map.of(), null, null, SemanticExpectedCardinality.UNKNOWN, false, List.of());
        }
        public SemanticGraphNode withAuthorization(AuthorizationPredicate p) {
            return new SemanticGraphNode(id, entityId, viaRelationship, viaConnection, parentId, p, fields, semanticAnnotations, traversalOrigin, intentOrigin, expectedCardinality, nullablePath, semanticConstraints);
        }
    }

    public record SemanticTraversalOrigin(EntityId sourceEntity, List<RelationshipId> path) {
        public SemanticTraversalOrigin { path = path == null ? List.of() : List.copyOf(path); }
    }
    public record SemanticIntentOrigin(String operation, String source) {}
    public enum SemanticExpectedCardinality { UNKNOWN, ONE, MANY }
}
